deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/layers/Deconvolution2D.java

```diff
@@ -0,0 +1,344 @@
+package org.deeplearning4j.nn.conf.layers;
+
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+import lombok.NoArgsConstructor;
+import lombok.ToString;
+import org.deeplearning4j.nn.api.Layer;
+import org.deeplearning4j.nn.api.ParamInitializer;
+import org.deeplearning4j.nn.conf.*;
+import org.deeplearning4j.nn.conf.distribution.Distribution;
+import org.deeplearning4j.nn.conf.inputs.InputType;
+import org.deeplearning4j.nn.layers.convolution.Deconvolution2DLayer;
+import org.deeplearning4j.nn.params.DeconvolutionParamInitializer;
+import org.deeplearning4j.nn.weights.WeightInit;
+import org.deeplearning4j.optimize.api.IterationListener;
+import org.deeplearning4j.util.ConvolutionUtils;
+import org.nd4j.linalg.activations.Activation;
+import org.nd4j.linalg.activations.IActivation;
+import org.nd4j.linalg.api.ndarray.INDArray;
+
+import java.util.Collection;
+import java.util.Map;
+
+/**
+ * 2D deconvolution layer configuration
+ *
+ * Deconvolutions are also known as transpose convolutions or fractionally strided convolutions.
+ * In essence, deconvolutions swap forward and backward pass with regular 2D convolutions.
+ *
+ * See the paper by Matt Zeiler for details:
+ * http://www.matthewzeiler.com/wp-content/uploads/2017/07/cvpr2010.pdf
+ *
+ * For an intuitive guide to convolution arithmetic and shapes, see:
+ * https://arxiv.org/abs/1603.07285v1
+ *
+ * @author Max Pumperla
+ */
+@Data
+@NoArgsConstructor
+@ToString(callSuper = true)
+@EqualsAndHashCode(callSuper = true)
+public class Deconvolution2D extends ConvolutionLayer {
+
+    /**
+     * Deconvolution2D layer
+     * nIn in the input layer is the number of channels
+     * nOut is the number of filters to be used in the net or in other words the depth
+     * The builder specifies the filter/kernel size, the stride and padding
+     * The pooling layer takes the kernel size
+     */
+    protected Deconvolution2D(BaseConvBuilder<?> builder) {
+        super(builder);
+        this.hasBias = builder.hasBias;
+        this.convolutionMode = builder.convolutionMode;
+        this.dilation = builder.dilation;
+        if (builder.kernelSize.length != 2)
+            throw new IllegalArgumentException("Kernel size of should be rows x columns (a 2d array)");
+        this.kernelSize = builder.kernelSize;
+        if (builder.stride.length != 2)
+            throw new IllegalArgumentException("Stride should include stride for rows and columns (a 2d array)");
+        this.stride = builder.stride;
+        if (builder.padding.length != 2)
+            throw new IllegalArgumentException("Padding should include padding for rows and columns (a 2d array)");
+        this.padding = builder.padding;
+        this.cudnnAlgoMode = builder.cudnnAlgoMode;
+        this.cudnnFwdAlgo = builder.cudnnFwdAlgo;
+        this.cudnnBwdFilterAlgo = builder.cudnnBwdFilterAlgo;
+        this.cudnnBwdDataAlgo = builder.cudnnBwdDataAlgo;
```

Comment:
All of these builder args (like, last 16 lines) should be handled in superclass constructor already... was there a reason for this?